### DISCUSSION TITLE
Best practices: what are the use cases for SWR vs getStaticProps?

### DISCUSSION THREAD

**Discussion by timcouchoud** [vercel/next.js#21100] [Help]
I am looking for best pratices using Next.js:

1. In which use cases should I use SWR? What are the pro and cons depending on each use case?
2. In which use cases should I use getStaticProps? What are the pro and cons depending on each use case?
3. In which use cases should I combine SWR and getStaticProps? What are the pro and cons depending on each use case?

A synoptic matrix would be a great thing, but maybe I am asking too much!:)

Thanks!

**Comment by stigkj** [comment]
SWR is for client side fetches, while getStaticProps is run on the server. Typically you should use SWR on the client side for fetching user specific data for a page that has been statically generated with data from getStaticProps.

**Comment by kvnwolf** [comment]
You can use three different strategies to fetch data on a Next.js application, you can either use a`fetch-then-build-cache-and-render`, `fetch-then-render` or `render-then-fetch` approach. I will explain them above:

### 1. `fetch-then-build-cache-and-render`

This fetches the page data at **build time**, using **Static Site Generation (SSG)** via `getStaticProps`, passes those props to the page component, builds the page, and then pushes it to a CDN. Think of it as a way to create **public** pages where the content will be the same for all viewers of your website (i.e. a blog post, a FAQ section, a documentation page). If you opt-in to **[Incremental Static Regeneration (ISR)](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration)**, the page will be re-generated and then pushed to the CDN, so next visitors will have fresh data.

### 2. `fetch-then-render`

This fetches the page data at **run time**, using **Server Side Rendering (SSR)** via `getServerSideProps`, passes those props to the page component, and renders its contents, only after the data has been fetched. Since the page can be different depending on the viewer (a settings page, a dashboard), it does not make sense to cache its result.

### 3. `render-then-fetch`

Just as the previous approach, fetches the page data at **run time** on the client using **Client Side Rendering (CSR)**, the only difference is that the request won't be blocked while fetching the data, so you can statically generate a page with a loading skeleton, push it to the CDN, render it to the client and then fetch the data when the page component mounts, (either using `swr`, or anything you prefer). This is my personal favorite over `fetch-then-render` because the page load will be faster, it will show the loading state and then re-render with the fetched data.

**Reply by kvnwolf** [reply]
You can also see this video by @leerob, which explain it better: https://www.youtube.com/watch?v=eMwTEo6AjDc
